File: LanguagesCognition/merge_sort.hs
mergesort'merge :: (Ord a) => [a] -> [a] -> [a]
mergesort'merge [] xs = xs
mergesort'merge xs [] = xs
mergesort'merge (x:xs) (y:ys)
    | (x < y) = x:mergesort'merge xs (y:ys)
    | otherwise = y:mergesort'merge (x:xs) ys
 
mergesort'splitinhalf :: [a] -> ([a], [a])
mergesort'splitinhalf xs = (take n xs, drop n xs)
    where n = (length xs) `div` 2 
 
mergesort :: (Ord a) => [a] -> [a]
mergesort xs 
    | (length xs) > 1 = mergesort'merge (mergesort ls) (mergesort rs)
    | otherwise = xs
    where (ls, rs) = mergesort'splitinhalf xs

readLines :: FilePath -> IO [String]
readLines = fmap lines . readFile
makeInteger :: [String] -> [Int]
makeInteger = map read
makeString :: [Int] -> [String]
makeString = map show

main :: IO ()
main = do
    putStrLn "Haskell"
    contents <- readLines "not_sorted.txt"
    let lst = makeInteger contents
    putStrLn "Sorting..."
    let sorted_lst = mergesort lst

    let outFileName = "sorted_haskell.txt"
    writeFile outFileName $ unlines (makeString sorted_lst)
    putStrLn "Done!"
    

File: LanguagesCognition/merge_sort.py
from collections import deque


def merge_sort(lst):
    if len(lst) <= 1:
        return lst

    def merge(left, right):
        merged, left, right = deque(), deque(left), deque(right)
        while left and right:
            merged.append(left.popleft() if left[0] <= right[0] else right.popleft())
        merged.extend(right if right else left)
        return list(merged)

    mid = int(len(lst) // 2)
    left = merge_sort(lst[:mid])
    right = merge_sort(lst[mid:])
    return merge(left, right)


if __name__ == "__main__":
    print("Python")
    with open('not_sorted.txt') as inf:
        nums = inf.readlines()
    lst = [int(x.strip()) for x in nums]

    # Prompt user
    print("Sorting...")

    # Merge Sort
    sorted_lst = merge_sort(lst)

    with open('sorted_python.txt', 'w+') as outf:
        for x in sorted_lst:
            outf.write(str(x) + '\n')

    # Prompt user
    print("Done!")

